### ISSUE TITLE
Limited stock discounted items, when bought, make it impossible to buy the regular-price item

### ISSUE THREAD

**Issue by dessysalta** [tgstation/tgstation#73560]
Reporting client version: 514.1587

## Round ID:
[200461](https://scrubby.melonmesa.com/round/200461)

## Reproduction:

Load into game as traitor
Open uplink and buy discounted item
Go to the same item (this time its full-price variant) and fail to buy it

Feels like an oversight, like the limited stock should only apply to the discount--because you can buy as many of an item as you want if you have the TC to get it at full price (so you can buy 5 chameleon kits at 2 TC, but can only buy one cham kit at 1 TC and then be barred from even the full-price version for the rest of the game, which makes no sense)

**Comment by Profakos** [comment]
Well, this is tricky. Before I have put discounts to be limited to 1, when they were infinite due to a bug that has been in the code for a year, they did an inverse of this: allowing you to buy infinite of things which has been limited in regular stock.

Maybe we can set up the shared stock connection to be one way?

**Comment by dessysalta** [comment]
I think the discounted items should be considered "separate" from the main list itself. So, a discounted item is a different listing and isn't considered the normal item in question; that way buying limited stock of that discount still lets you buy the normal price variant, and just locks out the discounted version.

Also, there's no real issue with buying two crab phones. Whether it's just you or three separate traitors with the same idea in mind, it's gonna happen. I really doubt one-off two-timers like that will make much of a difference if at all.

**Comment by Jacquerel** [comment]
I wonder if perhaps discounts should simply be refactored to work differently under the hood in order to kill two birds (this and #28286) with one stone. 
If the discount is just a modifier applied to the item which applies until the item is purchased and not a separate listing at all, this would solve both issues.

**Comment by Profakos** [comment]
My theory is that they were made a separate item so they can have their own tab.

**Comment by Profakos** [comment]
I think the best solution would be removing the discount item entries, and instead store them as separate discount datums. Then you can make each item display "'next x purchases are x percent off!" or "all sales today are x percent off!", and also add a filter button by the search bar that shows only discounted items. This would also slve the issue of buying the non discounted item by accident.
I suppose  that items with limited stocks would have to be exempt from the unlimited discounts, as it would be waste of an unlimited stock. Same for items that has less amounts for sale than the selected discount stock. What do you all think?

**Comment by dessysalta** [comment]
> its less about the crab phone and more about any possible future item that can be discounted but that the maker wouldn't want any one person to have two of, or being able to handle a possible future item thats either balanced around some kind of shared stock thing (which is possible now) or a possible future item that is balanced around only being able to buy two of them or something

Isn't it pretty obvious most items are balanced around buying only one of them? It's why getting two hand teleporters/eswords/revolvers/etc. makes it a curb stomp for you. We can argue all day about what item is intended to be used on its own and when, but most of _those_ issues stem from progtot and TC gain as a whole. Eswords/revolvers and other items were only ever intended to be bought _once_, maaaybe twice if you got lucky. If you wanna make everything in the uplink only buyable once (or implement stock for some items) be my guest, but I see ogtot coming back before that happens.

**tl;dr** You can buy a bajillion items if you have the TC as is, regardless of their strength. Blame progtot for that.

**Comment by Profakos** [comment]
I have started my work on my refactor. Firstly, I have nuked the "extra purchaseables" segment from the uplink, as it is no longer needed with the new discounts. And also, it is described as a list of items you can buy even if your uplink had different flags, but, when generating discounts, the generated items were always restricted by flags anyways.

**Comment by SinfulBliss** [comment]
Still an issue, bought the 1 TC discounted makarov armor piercing ammo, then was prevented from buying any more because it greyed out both the discount and the normal item.

I think it has something to do with whatever PR made it so you couldn't buy multiple discounts of the same item if someone could attach that.

**Comment by Profakos** [comment]
I could never finish my earlier refactor, due to lack of time and burnout, sadly. I am on vacation soon, so I could retry.

**Comment by Watermelon914** [comment]
> I could never finish my earlier refactor, due to lack of time and burnout, sadly. I am on vacation soon, so I could retry. Interim solution would be making discounts have a different Stock Key.

that would be the correct solution, I'll pr this in a bit

**Comment by Profakos** [comment]
Thank you.

Of course, this still wouldn't be perfect. With different stocks, you will be able to buy a second crab phone. But this is a small price to pay until a real refactor that turns discounts into datums instead of separate items.

**Comment by Watermelon914** [comment]
> Thank you.
> 
> Of course, this still wouldn't be perfect. With different stocks, you will be able to buy a second crab phone. But this is a small price to pay until a real refactor that turns discounts into datums instead of separate items.

What's the issue here? Is this not supposed to be possible?

**Comment by Profakos** [comment]
Hm, no hang on, your PR has probably fixed this issue. I will check in the future.

**Comment by Watermelon914** [comment]
My PR just gives each stock item its own stock. This means the original item and the crab phone will have different stocks, for example.
This will let you buy 2 crab phones if you buy the discounted version and the non-discounted version.

**Comment by Profakos** [comment]
The list is indeed called "Extra buyables" codewise, so I will concede on this one. 👍
